### PR TITLE
[QUARKS-124] [REVIEW] add Deadtime filter

### PR DESCRIPTION
**Pull request by dlaboss**
TODO Deadtime fvt, add support of PeriodicMXBean for control

### CODE REVIEW

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+ * </p><p>
+ * The deadtime period may be changed while the topology is running
+ * via {@link #setPeriod(long, TimeUnit)}.
+ * </p>
+ *
+ * @param <T> tuple type
+ */
+public class Deadtime<T> implements Predicate<T> {
+    private static final long serialVersionUID = 1L;
+    private transient long deadtimePeriod;
```
Comment (ddebrunner) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java]: Why are these fields transient?

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+     *        tuples received after the last passed tuple.
+     *        Specify a value of 0 to pass all received tuples.
+     *        Must be >= 0.
+     *        A period of 0 is used if the specified period is less than 1ms.
+     * @param unit {@link TimeUnit} of {@code deadtimePeriod}
+     */
+    public synchronized void setPeriod(long deadtimePeriod, TimeUnit unit) {
+        if (deadtimePeriod < 0)
+            throw new IllegalArgumentException("deadtimePeriod");
+        Objects.requireNonNull(unit, "unit");
```
Comment (ddebrunner) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java]: technically not needed

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+ * via {@link #setPeriod(long, TimeUnit)}.
+ * </p>
+ *
+ * @param <T> tuple type
+ */
+public class Deadtime<T> implements Predicate<T> {
+    private static final long serialVersionUID = 1L;
+    private transient long deadtimePeriod;
+    private transient TimeUnit deadtimeUnit;
+    private transient long deadtimePeriodMillis;
```
Comment (ddebrunner) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java]: the same info is being stored twice, it seems to be only for the toString() method, is it worth it?

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java [hunk 4]
```diff
@@ -0,0 +1,98 @@
+package quarks.analytics.sensors;
+
+import java.util.Date;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+import quarks.function.Predicate;
+
+/**
+ * A generic "deadtime" {@link Predicate}.
```
Comment (ddebrunner) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java]: Saw this with another class, why is "deadtime" in quotes?

Maybe:

Predicate that discards values for a period of time after passing a value.

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    }
+
+    @Override
+    public boolean test(T value) {
+        long now = System.currentTimeMillis(); 
+        if (now < nextPassTimeMillis)
+            return false;
+        else {
+            lastPassTimeMillis = now;
+            nextPassTimeMillis = lastPassTimeMillis + deadtimePeriodMillis;
```
Comment (ddebrunner) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java]: might be clearer (and faster) it this was:

` nextPassTimeMillis = now + deadtimePeriodMillis;`

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Filters.java [hunk 6]
```diff
@@ -114,4 +114,27 @@ private Filters() {}
 
         return stream.filter(new Deadband<>(value, inBand));
     }
+    
+    /**
+     * A filter that discards tuples received during the "deadtime period"
```
Comment (ddebrunner) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Filters.java]: Not English! and too long for the summary that is seen in Javadoc.

Maybe just:

Deadtime filter.

(c.f. Deadband filter)

Then have the following sentences describe what it does.

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Filters.java [hunk 7]
```diff
@@ -114,4 +114,27 @@ private Filters() {}
...
+     * following a tuple this is allowed to pass through.
+     * <p>
+     * E.g., for a deadtime period of 30 minutes, after letting a tuple
+     * pass through, any tuples received during the next 30 minutes are
+     * filtered out.  Then the next arriving tuple is passed through and
+     * a new deadtime period is begun.
+     * </p><p>
+     * See {@link Deadtime} for information about changing the deadtime period
```
Comment (ddebrunner) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Filters.java]: Is this for the future, as there's no mechanism for the caller of this method to change the period?

Comment (dlaboss) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Filters.java]: reworked

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Filters.java (hunk 6 above)
Comment (dlaboss) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Filters.java]: reworked

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java (hunk 4 above)
Comment (dlaboss) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java]: reworked

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java (hunk 5 above)
Comment (dlaboss) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java]: reworked

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java (hunk 2 above)
Comment (dlaboss) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java]: Using requireNonNull with a msg is better for the caller than just receiving a "naked" NPE a few lines down, right?  Using it also enforces the requirement even if later the code below it doesn't end up referencing ``unit``.

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java (hunk 3 above)
Comment (dlaboss) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java]: I'll toss deadtime{Period,Unit}

File: analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java (hunk 1 above)
Comment (dlaboss) [inline analytics/sensors/src/main/java/quarks/analytics/sensors/Deadtime.java]: thanks!  meant to type volatile.

**Pull request merged**